File: CODE/PYTHON/LIBRARY/processing.py

# -- IMPORTS

import base64;
import hashlib;

# -- FUNCTIONS


def get_text_hash(
    text
    ):

    return hashlib.md5( text.encode( "utf8" ) ).hexdigest();


# ~~


def get_text_uuid(
    text
    ):

    if ( text == "" ):

        return "";

    else:

        hash = hashlib.md5( text.encode( "utf8" ) ).hexdigest();

        return (
            hash[ 0:8 ]
            + "-"
            + hash[ 8:12 ]
            + "-"
            + hash[ 12:16 ]
            + "-"
            + hash[ 16:20 ]
            + "-"
            + hash[ 20:32 ]
            );


# ~~


def get_text_tuid(
    text
    ):

    if ( text == "" ):

        return "";

    else:

        hash = hashlib.md5( text.encode( "utf8" ) ).hexdigest();
        byte_array = [ None ] * 16;

        character_index = 0;

        while ( character_index < 32 ):

            byte_array[ character_index >> 1 ] = chr( int( hash[ character_index:( character_index + 2 ) ], 16 ) );
            character_index += 2;

        tuid = base64.b64encode( "".join( byte_array ).encode( "latin1" ) ).decode( "ascii" );

        return (
            tuid
                .replace( "+", "-" )
                .replace( "/", "_" )
                .replace( "=", "" )
            );


# ~~


def process_primed_text(
    primed_text
    ):

    if ( primed_text.startswith( "‼#" ) ):

        return get_text_uuid( primed_text[ 2: ] );

    elif ( primed_text.startswith( "‼%" ) ):

        return get_text_tuid( primed_text[ 2: ] );

    else:

        return primed_text;

File: CODE/TYPESCRIPT/TEXT/processing.ts
// -- IMPORTS

import md5 from "md5";

// -- FUNCTIONS

export function getTextHash(
    text: string
    )
    : string
{
    return md5( text );
}

// ~~

export function getTextUuid(
    text: string
    )
    : string
{
    if ( text === "" )
    {
        return "";
    }
    else
    {
        let hash = md5( text );

        return (
            hash.slice( 0, 8 )
            + "-"
            + hash.slice( 8, 12 )
            + "-"
            + hash.slice( 12, 16 )
            + "-"
            + hash.slice( 16, 20 )
            + "-"
            + hash.slice( 20, 32 )
            );
    }
}

// ~~

export function getTextTuid(
    text: string
    )
    : string
{
    if ( text === "" )
    {
        return "";
    }
    else
    {
        let hash = md5( text );

        let byteArray = new Array( 16 );

        for ( let characterIndex = 0;
              characterIndex < 32;
              characterIndex += 2 )
        {
            byteArray[ characterIndex >> 1 ] = String.fromCharCode( parseInt( hash.slice( characterIndex, characterIndex + 2 ), 16 ) );
        }

        let tuid = btoa( byteArray.join( "" ) );

        return (
            tuid
                .replaceAll( "+", "-" )
                .replaceAll( "/", "_" )
                .replaceAll( "=", "" )
            );
    }
}

// ~~

export function processPrimedText(
    primedText: string
    )
    : string
{
    if ( primedText.startsWith( "‼#" ) )
    {
        return getTextUuid( primedText.slice( 2 ) );
    }
    else if ( primedText.startsWith( "‼%" ) )
    {
        return getTextTuid( primedText.slice( 2 ) );
    }
    else
    {
        return primedText;
    }
}
